File: src/nit/git.py
from __future__ import annotations

import logging
import subprocess
from pathlib import Path

logger = logging.getLogger(__name__)

GIT_TIMEOUT = 30


def _run(args: list[str], cwd: Path | None = None) -> str:
    logger.debug("Running: %s", " ".join(args))
    try:
        result = subprocess.run(
            args,
            capture_output=True,
            text=True,
            cwd=cwd,
            timeout=GIT_TIMEOUT,
        )
    except subprocess.TimeoutExpired:
        logger.warning("Git command timed out: %s", " ".join(args))
        raise subprocess.CalledProcessError(1, args, "", "Command timed out")
    if result.returncode != 0:
        raise subprocess.CalledProcessError(
            result.returncode,
            args,
            result.stdout,
            result.stderr,
        )
    return result.stdout


def get_repo_root(cwd: Path | None = None) -> Path:
    out = _run(["git", "rev-parse", "--show-toplevel"], cwd=cwd)
    return Path(out.strip())


def get_current_branch(cwd: Path | None = None) -> str:
    return _run(["git", "branch", "--show-current"], cwd=cwd).strip()


def get_main_branch(cwd: Path | None = None) -> str:
    for name in ("main", "master"):
        for ref in (f"refs/remotes/origin/{name}", f"refs/heads/{name}"):
            result = subprocess.run(
                ["git", "rev-parse", "--verify", ref],
                capture_output=True,
                text=True,
                cwd=cwd,
                timeout=GIT_TIMEOUT,
            )
            if result.returncode == 0:
                return f"origin/{name}" if ref.startswith("refs/remotes") else name
    return "main"


def get_merge_base(base: str, cwd: Path | None = None) -> str:
    return _run(["git", "merge-base", base, "HEAD"], cwd=cwd).strip()


def _append_path_filter(cmd: list[str], path_filter: str | None) -> list[str]:
    if path_filter:
        return cmd + ["--", path_filter]
    return cmd


def _append_whitespace_flag(cmd: list[str], ignore_whitespace: bool) -> list[str]:
    if ignore_whitespace:
        cmd = cmd + ["-w"]
    return cmd


def _build_diff_cmd(
    base_cmd: list[str], ignore_whitespace: bool, path_filter: str | None
) -> list[str]:
    cmd = _append_whitespace_flag(base_cmd, ignore_whitespace)
    return _append_path_filter(cmd, path_filter)


def get_branch_diff(
    cwd: Path | None = None,
    path_filter: str | None = None,
    ignore_whitespace: bool = False,
) -> str:
    base = get_main_branch(cwd)
    cmd = _build_diff_cmd(["git", "diff", f"{base}...HEAD"], ignore_whitespace, path_filter)
    return _run(cmd, cwd=cwd)


def get_unstaged_diff(
    cwd: Path | None = None,
    path_filter: str | None = None,
    ignore_whitespace: bool = False,
) -> str:
    cmd = _build_diff_cmd(["git", "diff"], ignore_whitespace, path_filter)
    return _run(cmd, cwd=cwd)


def get_all_uncommitted_diff(
    cwd: Path | None = None,
    path_filter: str | None = None,
    ignore_whitespace: bool = False,
) -> str:
    cmd = _build_diff_cmd(["git", "diff", "HEAD"], ignore_whitespace, path_filter)
    return _run(cmd, cwd=cwd)


def get_staged_diff(
    cwd: Path | None = None,
    path_filter: str | None = None,
    ignore_whitespace: bool = False,
) -> str:
    cmd = _build_diff_cmd(["git", "diff", "--cached"], ignore_whitespace, path_filter)
    return _run(cmd, cwd=cwd)


def apply_patch(
    patch_text: str, cwd: Path | None = None, cached: bool = False, reverse: bool = False
) -> str:
    cmd = ["git", "apply"]
    if cached:
        cmd.append("--cached")
    if reverse:
        cmd.append("--reverse")
    logger.debug("Running: %s (with stdin patch)", " ".join(cmd))
    try:
        result = subprocess.run(
            cmd,
            input=patch_text,
            capture_output=True,
            text=True,
            cwd=cwd,
            timeout=GIT_TIMEOUT,
        )
    except subprocess.TimeoutExpired:
        raise subprocess.CalledProcessError(1, cmd, "", "Command timed out")
    if result.returncode != 0:
        raise subprocess.CalledProcessError(result.returncode, cmd, result.stdout, result.stderr)
    return result.stdout


def commit(message: str, cwd: Path | None = None) -> str:
    return _run(["git", "commit", "-m", message], cwd=cwd)


def get_commit_range_diff(
    commit_range: str,
    cwd: Path | None = None,
    path_filter: str | None = None,
    ignore_whitespace: bool = False,
) -> str:
    cmd = _build_diff_cmd(["git", "diff", commit_range], ignore_whitespace, path_filter)
    return _run(cmd, cwd=cwd)

File: internal/git/git.go
package git

import (
	"bytes"
	"context"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"
)

const GitTimeout = 30 * time.Second

// run executes a git command with timeout and returns stdout.
func run(args []string, cwd string) (string, error) {
	ctx, cancel := context.WithTimeout(context.Background(), GitTimeout)
	defer cancel()
	return runCtx(ctx, args, cwd, "")
}

// runWithStdin executes a git command with stdin input.
func runWithStdin(args []string, cwd string, stdin string) (string, error) {
	ctx, cancel := context.WithTimeout(context.Background(), GitTimeout)
	defer cancel()
	return runCtx(ctx, args, cwd, stdin)
}

func runCtx(ctx context.Context, args []string, cwd string, stdin string) (string, error) {
	cmd := exec.CommandContext(ctx, args[0], args[1:]...)
	if cwd != "" {
		cmd.Dir = cwd
	}
	var stdout, stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr
	if stdin != "" {
		cmd.Stdin = strings.NewReader(stdin)
	}
	err := cmd.Run()
	if ctx.Err() == context.DeadlineExceeded {
		return "", fmt.Errorf("git command timed out: %s", strings.Join(args, " "))
	}
	if err != nil {
		return "", fmt.Errorf("%s: %s", strings.Join(args, " "), strings.TrimSpace(stderr.String()))
	}
	return stdout.String(), nil
}

// GetRepoRoot returns the absolute path to the git repository root.
func GetRepoRoot(cwd string) (string, error) {
	out, err := run([]string{"git", "rev-parse", "--show-toplevel"}, cwd)
	if err != nil {
		return "", err
	}
	return strings.TrimSpace(out), nil
}

// GetCurrentBranch returns the current branch name, or "" if detached.
func GetCurrentBranch(cwd string) (string, error) {
	out, err := run([]string{"git", "branch", "--show-current"}, cwd)
	if err != nil {
		return "", err
	}
	return strings.TrimSpace(out), nil
}

// GetMainBranch returns the main/master branch name, preferring remote refs.
func GetMainBranch(cwd string) string {
	for _, name := range []string{"main", "master"} {
		for _, ref := range []string{
			fmt.Sprintf("refs/remotes/origin/%s", name),
			fmt.Sprintf("refs/heads/%s", name),
		} {
			cmd := exec.Command("git", "rev-parse", "--verify", ref)
			if cwd != "" {
				cmd.Dir = cwd
			}
			cmd.Stdout = nil
			cmd.Stderr = nil
			if cmd.Run() == nil {
				if strings.HasPrefix(ref, "refs/remotes") {
					return "origin/" + name
				}
				return name
			}
		}
	}
	return "main"
}

// GetMergeBase returns the merge-base commit SHA between base and HEAD.
func GetMergeBase(base, cwd string) (string, error) {
	out, err := run([]string{"git", "merge-base", base, "HEAD"}, cwd)
	if err != nil {
		return "", err
	}
	return strings.TrimSpace(out), nil
}

func appendWhitespaceFlag(cmd []string, ignoreWhitespace bool) []string {
	if ignoreWhitespace {
		return append(cmd, "-w")
	}
	return cmd
}

func appendPathFilter(cmd []string, pathFilter string) []string {
	if pathFilter != "" {
		return append(cmd, "--", pathFilter)
	}
	return cmd
}

func buildDiffCmd(baseCmd []string, ignoreWhitespace bool, pathFilter string) []string {
	cmd := appendWhitespaceFlag(baseCmd, ignoreWhitespace)
	return appendPathFilter(cmd, pathFilter)
}

// GetBranchDiff returns the diff between the main branch and HEAD.
func GetBranchDiff(cwd, pathFilter string, ignoreWhitespace bool) (string, error) {
	base := GetMainBranch(cwd)
	return GetBranchDiffWithBase(base, cwd, pathFilter, ignoreWhitespace)
}

// GetBranchDiffWithBase returns the diff using a pre-resolved base branch.
func GetBranchDiffWithBase(base, cwd, pathFilter string, ignoreWhitespace bool) (string, error) {
	cmd := buildDiffCmd([]string{"git", "diff", base + "...HEAD"}, ignoreWhitespace, pathFilter)
	return run(cmd, cwd)
}

// GetDiffStat returns a compact stat summary for change detection.
// Much cheaper than full diff output — used for auto-refresh debouncing.
func GetDiffStat(cwd string, args ...string) (string, error) {
	cmd := append([]string{"git", "diff", "--stat"}, args...)
	return run(cmd, cwd)
}

// GetUnstagedDiff returns the diff of unstaged changes.
func GetUnstagedDiff(cwd, pathFilter string, ignoreWhitespace bool) (string, error) {
	cmd := buildDiffCmd([]string{"git", "diff"}, ignoreWhitespace, pathFilter)
	return run(cmd, cwd)
}

// GetStagedDiff returns the diff of staged changes.
func GetStagedDiff(cwd, pathFilter string, ignoreWhitespace bool) (string, error) {
	cmd := buildDiffCmd([]string{"git", "diff", "--cached"}, ignoreWhitespace, pathFilter)
	return run(cmd, cwd)
}

// GetUpstreamRef returns the upstream tracking ref, or "" if not set.
func GetUpstreamRef(cwd string) string {
	out, err := run([]string{"git", "rev-parse", "--abbrev-ref", "@{upstream}"}, cwd)
	if err != nil {
		return ""
	}
	return strings.TrimSpace(out)
}

// GetUnpushedDiff returns the diff between the upstream tracking branch and HEAD.
func GetUnpushedDiff(cwd, pathFilter string, ignoreWhitespace bool) (string, error) {
	upstream := GetUpstreamRef(cwd)
	if upstream == "" {
		return "", fmt.Errorf("no upstream branch set")
	}
	cmd := buildDiffCmd([]string{"git", "diff", upstream}, ignoreWhitespace, pathFilter)
	return run(cmd, cwd)
}

// GetCommitRangeDiff returns the diff for an arbitrary commit range.
func GetCommitRangeDiff(commitRange, cwd, pathFilter string, ignoreWhitespace bool) (string, error) {
	cmd := buildDiffCmd([]string{"git", "diff", commitRange}, ignoreWhitespace, pathFilter)
	return run(cmd, cwd)
}

// ApplyPatch applies a patch via git apply.
func ApplyPatch(patchText, cwd string, cached, reverse bool) (string, error) {
	cmd := []string{"git", "apply"}
	if cached {
		cmd = append(cmd, "--cached")
	}
	if reverse {
		cmd = append(cmd, "--reverse")
	}
	return runWithStdin(cmd, cwd, patchText)
}

// Commit creates a commit with the given message.
func Commit(message, cwd string) (string, error) {
	return run([]string{"git", "commit", "-m", message}, cwd)
}

// GetUntrackedFiles returns a list of untracked file paths.
func GetUntrackedFiles(cwd string) ([]string, error) {
	out, err := run([]string{"git", "ls-files", "--others", "--exclude-standard"}, cwd)
	if err != nil {
		return nil, err
	}
	var files []string
	for _, f := range strings.Split(strings.TrimSpace(out), "\n") {
		if f != "" {
			files = append(files, f)
		}
	}
	return files, nil
}

// isBinaryFile checks if a file appears to be binary by looking for null bytes.
func isBinaryFile(path string) bool {
	f, err := os.Open(path)
	if err != nil {
		return false
	}
	defer func() { _ = f.Close() }()
	buf := make([]byte, 8192)
	n, _ := f.Read(buf)
	return bytes.Contains(buf[:n], []byte{0})
}

// untrackedFileDiff generates a unified diff for an untracked file as if newly added.
func untrackedFileDiff(path, cwd string) string {
	dir := cwd
	if dir == "" {
		dir, _ = os.Getwd()
	}
	fullPath := filepath.Join(dir, path)

	// Skip binary files — emit a marker instead of reading content
	if isBinaryFile(fullPath) {
		return fmt.Sprintf("diff --git a/%s b/%s\nBinary files /dev/null and b/%s differ\n", path, path, path)
	}

	data, err := os.ReadFile(fullPath)
	if err != nil {
		return ""
	}
	content := string(data)
	lines := strings.Split(content, "\n")
	if len(lines) > 0 && lines[len(lines)-1] == "" {
		lines = lines[:len(lines)-1]
	}
	n := len(lines)

	var b strings.Builder
	fmt.Fprintf(&b, "diff --git a/%s b/%s\n", path, path)
	b.WriteString("new file mode 100644\n")
	b.WriteString("--- /dev/null\n")
	fmt.Fprintf(&b, "+++ b/%s\n", path)
	fmt.Fprintf(&b, "@@ -0,0 +1,%d @@\n", n)
	for _, line := range lines {
		fmt.Fprintf(&b, "+%s\n", line)
	}
	return b.String()
}

// GetUntrackedDiff returns synthetic diff text for all untracked files.
func GetUntrackedDiff(cwd, pathFilter string) (string, error) {
	files, err := GetUntrackedFiles(cwd)
	if err != nil {
		return "", err
	}
	var b strings.Builder
	for _, f := range files {
		if pathFilter != "" && !strings.HasPrefix(f, pathFilter) && f != pathFilter {
			continue
		}
		b.WriteString(untrackedFileDiff(f, cwd))
	}
	return b.String(), nil
}
